### PR TITLE
Link custom metric

### PR DESCRIPTION
**Pull request by caspervdw**
This reimplements the kwarg `neighbor_strategy`, but the BTree is now based on the BallTree from scikit-learn, allowing for custom metrics.

To define a custom metric, you can:
 - pass a function as `to_eucl` that converts an `(N, ndim)` array of coordinates to an `(N, ndim2)` array of Euclidean coordinates (fast)
 - pass a function as `dist_func` that takes two `(ndim,)` arrays of positions and returns a distance (`float`). See [here](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.DistanceMetric.html)

I had to change the default rescaling of coordinates with `search_range`, that is why this PR is rather extensive. We have some (maybe negligible) performance increase for isotropic search ranges, because coordinates are not rescaled.

### CODE REVIEW

**Comment by caspervdw** [comment]
@tacaswell @danielballan @nkeim Bumping for the release of v0.4

File: trackpy/linking/linking.py [hunk 1]
```diff
@@ -96,15 +92,23 @@ def link(f, search_range, pos_columns=None, t_column='frame', **kwargs):
         algorithm used to resolve subnetworks of nearby particles
         'auto' uses numba if available
         'drop' causes particles in subnetworks to go unlinked
+    neighbor_strategy : {'KDTree', 'BTree'}
+        algorithm used to identify nearby features. Default 'KDTree'.
+    to_eucl : function, optional
```
Comment (nkeim) [inline trackpy/linking/linking.py]: Is this something that could be done by just processing the data before linking? As opposed to `dist_func`, which needs to be incorporated here. I am just thinking about our trajectory with data pre-processing in pims.

File: trackpy/linking/subnet.py [hunk 2]
```diff
@@ -106,35 +89,147 @@ def coords_df(self):
...
+    @property
+    def coords_mapped(self):
+        if not self._clean:
+            self.rebuild()
+        if self._btree is None:
+            return np.empty((0, self.ndim))
+        else:
+            return self._btree.data
 
+    def rebuild(self):
```
Comment (nkeim) [inline trackpy/linking/subnet.py]: So great that prediction can work with this new code!

File: trackpy/tests/test_linking.py [hunk 3]
```diff
@@ -653,9 +697,50 @@ def test_drop_link(self):
...
+        # link in normal (2D Euclidean) coordinates
+        expected = self.link(f, search_range)
+
+        # compute radial coordinates
+        f_radial = f.copy()
+        f_radial['angle'] = np.arctan2(f_radial['y'], f_radial['x'])
+        f_radial['r'] = np.sqrt(f_radial['y']**2 + f_radial['x']**2)
+        # leave x, y for the comparison at the end
+
+        def dist_func(a, b):
```
Comment (nkeim) [inline trackpy/tests/test_linking.py]: This is monumentally awesome.

Comment (nkeim) [inline trackpy/tests/test_linking.py]: Upon merge, we should create an issue to write a tutorial or a short example about how to do this.

**Comment by nkeim** [comment]
As expected, no performance regression.

File: trackpy/linking/linking.py (hunk 1 above)
Comment (caspervdw) [inline trackpy/linking/linking.py]: Yes this `to_eucl` might be done as preprosessing as well. We do need the implementation anyway for the anisotropic-pixel size rescaling. Might as well expose it to the user.

File: trackpy/tests/test_linking.py (hunk 3 above)
Comment (caspervdw) [inline trackpy/tests/test_linking.py]: Yes this really would have helped me in my PhD! And a tutorial would be great to have.

**Comment by caspervdw** [comment]
@nkeim I will have some time to push the release of v0.4 during the next couple of weeks only, could you please have another look so that we can finish this? I guess #445 can wait for a further release.

**Comment by nkeim** [comment]
@caspervdw Thanks. I don't see anything significant that still needs addressing, but I will get back to you by Monday at the latest. (Classes just started here.)
